tests/simulator_test.cpp

```diff
@@ -2,82 +2,217 @@
-  Ray tiltedFromXY(kVZero, Vec3(2 * manager.collectorRadius(),
-                                2 * manager.collectorRadius(), 2));
-  ASSERT_FALSE(performHitAtEnergyCollectors(tiltedFromXY, &hitData));
-
-  Ray straightUpMoved(
-      Vec3(2 * manager.collectorRadius(), 2 * manager.collectorRadius(), 0),
-      Vec3(0, 0, 1));
-  ASSERT_FALSE(performHitAtEnergyCollectors(straightUpMoved, &hitData));
-} */
\ No newline at end of file
+  ASSERT_EQ(performHitCollector(energyCollectors, straightUp, &hitData), true);
+
+  const float collectorPositionRadius = 4;
+  const float collectorAngle = 2 * kPi / (numCollectors - 1);
+  const float refCollectorRadius =
+      collectorPositionRadius * std::sqrt(2 - 2 * std::cos(collectorAngle));
+
+  Vec3 collisionPointStraightUp =
+      Vec3(0, 0, collectorPositionRadius - refCollectorRadius);
+  ASSERT_EQ(hitData.collisionPoint(), collisionPointStraightUp);
+
+  Ray straightDown(kVecZero, -kVecUp);
+  ASSERT_FALSE(performHitCollector(energyCollectors, straightDown, &hitData))
+      << "Collision Point: " << hitData.collisionPoint();
+
+  Ray alongX(kVecZero, kVecX);
+  ASSERT_TRUE(performHitCollector(energyCollectors, alongX, &hitData));
+
+  ASSERT_EQ(alongX.at(collectorPositionRadius - refCollectorRadius),
+            hitData.collisionPoint());
+
+  Ray alongY(kVecZero, kVecY);
+  ASSERT_TRUE(performHitCollector(energyCollectors, alongY, &hitData));
+
+  ASSERT_EQ(alongY.at(collectorPositionRadius - refCollectorRadius),
+            hitData.collisionPoint());
+
+  Ray atSixty(kVecZero, Vec3(std::cos(deg2rad(60)), 0, std::sin(deg2rad(60))));
+  ASSERT_TRUE(performHitCollector(energyCollectors, atSixty, &hitData));
+
+  Ray atSixtyOther(kVecZero,
+                   Vec3(-std::cos(deg2rad(60)), 0, std::sin(deg2rad(60))));
+  ASSERT_TRUE(performHitCollector(energyCollectors, atSixtyOther, &hitData));
+
+  Ray atSixtyXY(kVecZero, Vec3(std::cos(deg2rad(60)), std::cos(deg2rad(60)),
+                               std::sin(deg2rad(60))));
+  ASSERT_FALSE(performHitCollector(energyCollectors, atSixtyXY, &hitData))
+      << "Collision Point: " << hitData.collisionPoint();
+
+  Ray atSixtyXYOther(kVecZero,
+                     Vec3(-std::cos(deg2rad(60)), -std::cos(deg2rad(60)),
+                          std::sin(deg2rad(60))));
+  ASSERT_FALSE(performHitCollector(energyCollectors, atSixtyXYOther, &hitData))
+      << "Collision Point: " << hitData.collisionPoint();
+}
+
+TEST_F(EnergyCollectorTest, EvenNumOfEnergyCollectorTest) {
+  const FakeModel nonEmptyModel(false);
+
+  const int numCollectors = 36;
+  auto energyCollectors = buildCollectors(nonEmptyModel, numCollectors);
+  ASSERT_EQ(energyCollectors.size(), numCollectors);
+
+  Ray straightUp(kVecZero, kVecUp);
+  RayHitData hitData;
+  ASSERT_TRUE(performHitCollector(energyCollectors, straightUp, &hitData));
+
+  const float collectorPositionRadius = 4;
+  const float collectorAngle = 2 * kPi / (numCollectors - 2);
+  const float refCollectorRadius =
```

Comment:
Isn't this just:

```
(OriginOfLastCollector + OriginOfPenultimateCollector).magnitude()
```

Not sure if this is a good idea though, as this would mean that it would depend on the order in which you are placing the collectors.

Reply:
Can you tell me number of line code, where you want to put it? Unfortunatly git doesnt display it clearly ;/

I used originOflastCollector and OriginOfPenultimateCollector to calculate hit point between two collectors only, because this was the easiest way to calculate it and probably the easiest way to explain via code. Actually I dont have any idea on how can I get this collision point without those two points.  

I know, that order of placing collectors should not matter, but this is very hard to archieve for this test, so I thought that because order of placing energy collectors in simulation doesnt matter, I will use it for my advantage. I changed the order of placing collectors at their desired positions, to support this kind of test.

Reply:
Yes, I understand. My point is that now you are calculating radius in one way, but you are using the distance between points below as a second way of getting the same thing.

Doing something internally in the implementation add support for some tests is a very bad practice, because it makes the code/tests brittle. Ideally you want to export something from the API that is giving you all the info you need.


edit: discussed offline that you can just go thought the collectors and find max Z and then subtract the height. Please leave the comment you have though (+ add explanation about the max Z) as it was actually helpful.

Reply:
now this test doesnt't depend on the order of collectors. Let me know if you think that I need to add more explaination to this ;)

Reply:
Prove it and shuffle the collectors in the impl.

Reply:
![image](https://user-images.githubusercontent.com/45213495/98830602-e4303600-243a-11eb-9c4d-de2dd789187c.png)

![image](https://user-images.githubusercontent.com/45213495/98830621-ebefda80-243a-11eb-85d1-282cc0f7da16.png)

works ;)